### ISSUE TITLE
BUG: Restricted Blocks to root level via parent "core/post-content" are nestable and groupable again

### ISSUE THREAD

**Issue by CreativeDive** [WordPress/gutenberg#49662]
### Description

Following @youknowriad's advice, I created a new issue.

I tested this behavior with WP 6.2 and Gutenberg 15.5.0 and it no longer works. If the block has "`core/post-content`" set in the "`parent`" attribute in `block.json`, the block can be grouped. But that shouldn't be possible.

I have other custom blocks that are restricted to the root level of the editor, but the options to group these blocks are available. But that shouldn't be possible.

It would be nice if it works as expected again.

It seems it's handled here:

https://github.com/WordPress/gutenberg/blob/a43afc673255965d919803cb32923a82855c1de4/packages/block-editor/src/store/selectors.js#L1511-L1525

I've made a simple test:

```javascript
blockTypeParent = blockType.parent !== undefined ? blockType.parent : null;

if( blockTypeParent ) {
    if( blockTypeParent.includes('core/post-content') ) {
        return false;
    }
}
```

In any case, a block has `core/post-content` defined in `parent`, my test comes up with `return false;`, which should disable the canInsert possibility. But it's still possible. Can someone help me to find the reason why it stopped working?

It used to work with #18260 #16411, but not anymore.

### Step-by-step reproduction instructions

1. Blocks with parent setting "`core/post-content`" are groupable
2. Blocks with parent setting "`core/post-content`" are nestable

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by CreativeDive** [comment]
In addition to this bug, a new feature would make perfect sense. To prevent a block from being nested, this is easy to implement with the `hasBlockSupport()` function.

Inside the Gutenberg `canInsertBlockTypeUnmemoized()` function we could add something like:

```javascript
if( ! hasBlockSupport(blockName, 'nestable', true) ) {
    return false;
}
```

if the `block.json` contains:

```json
"supports": {
    "nestable": false
}
```

**Comment by youknowriad** [comment]
I've tried to debug this, and I'm not sure that it's a regression. I don't see how we could check whether a "transform" will result in an "allowed parent" or not unless we actually run the transform function, and this is something that we never did AFAIK.

**Comment by CreativeDive** [comment]
Thank you for response @youknowriad. It seems if we extend the check by `rootClientId` this will work like expected.

```javascript
blockTypeParent = blockType.parent !== undefined ? blockType.parent : null;

if( rootClientId && blockTypeParent ) {
    if( blockTypeParent.includes('core/post-content') ) {
        return false;
    }
}
```

I've tested it and in this case the block is no longer nestable but we can insert the block on the root level.

Is this something we can add to improve the editor experience?

Edit: But it seems the block can still be grouped, which shouldn't work.

**Comment by CreativeDive** [comment]
@youknowriad I debugged the code further and found a solution.

### 1. Let's say we define a new function `blockCanBeNested()`:

```javascript
function blockCanBeNested( blocks, rootClientId ) {        

    let normalizedBlocks = Array.isArray(blocks) ? blocks : [blocks]; 
    let hasRestrictedBlock;
    
    normalizedBlocks.forEach( function( block ) {      
    
        var blockType = getBlockType( block.name );          
        var blockTypeParent = blockType.parent !== undefined ? blockType.parent : null;
        
        if( rootClientId && blockTypeParent ) {
            if( blockTypeParent.includes('core/post-content') ) {
                hasRestrictedBlock = true;
            }
        }
        
    });
                   
    if( hasRestrictedBlock ) {
        return false;
    }
    
    return true;    
    
}
```

### 2. In the `useConvertToGroupButtonProps()` function:

https://github.com/WordPress/gutenberg/blob/1d4e333988441854e532ed604ed870aef6be8f72/packages/block-editor/src/components/convert-to-group-buttons/use-convert-to-group-button-props.js#L33

We change ...

```javascript
const groupingBlockAvailable = canInsertBlockType( _groupingBlockName, rootClientId );

const _blocksSelection = getBlocksByClientId( _clientIds );
```

to ...

```javascript
const _blocksSelection = getBlocksByClientId( _clientIds );

const groupingBlockAvailable = blockCanBeNested(_blocksSelection, true) && canInsertBlockType( _groupingBlockName, rootClientId );
```

### 3. In the `canInsertBlockTypeUnmemoized` const:

https://github.com/WordPress/gutenberg/blob/a43afc673255965d919803cb32923a82855c1de4/packages/block-editor/src/store/selectors.js#L1511

We change ...

```javascript
if ( ! blockType ) {
    return false;
}
```

to ...

```javascript
if ( ! blockType ) {
    return false;
}

if( ! blockCanBeNested( blockType, rootClientId ) ) {
    return false;
}
```

The result ends finally in, the block type can no longer be nested and is no longer groupable, if the block.json has the parent `core/post-content` defined.

I've made a test with the `core/cover` block and it seems it's working like expected.

### Before:

![Apr-11-2023 10-43-54](https://user-images.githubusercontent.com/31563167/231105973-b70ad800-ff19-407f-ac51-6a8859a26a09.gif)

### After:

![Apr-11-2023 10-37-43](https://user-images.githubusercontent.com/31563167/231104548-5d26489f-365b-40ef-9829-7771a15b2000.gif)

**Comment by youknowriad** [comment]
What about the block switcher? (When you click the "block icon") Is the "grouping" action hidden there too?

**Comment by CreativeDive** [comment]
@youknowriad do you mean the block transformation options? These are still available, I think an additional condition for the block transform context can be solve that too. Should I see if I can find a solution?

**Comment by CreativeDive** [comment]
@youknowriad 

If we add an additional condition on the `getBlockTransformItems` const the block transformation is no longer possible.

https://github.com/WordPress/gutenberg/blob/a43afc673255965d919803cb32923a82855c1de4/packages/block-editor/src/store/selectors.js#L2101

Change ...

```javascript
const normalizedBlocks = Array.isArray( blocks ) ? blocks : [ blocks ];
```

to ...

```javascript
const normalizedBlocks = Array.isArray( blocks ) ? blocks : [ blocks ];
if( ! blockCanBeNested( blocks, true ) ) {    
   return [];        
}
```

Finally it seems that's the solution for the issue. Of course, this should be tested again by a core developer.

**Comment by youknowriad** [comment]
This is what I meant here https://github.com/WordPress/gutenberg/issues/49662#issuecomment-1501637996

Basically, the only solution is to actually run the transformation before hand and I'm not totally convinced that it's wise performance wise.

Your proposal is not correct because not all transformations "nest" the blocks.

**Comment by CreativeDive** [comment]
What can I say? I like to contribute and solve an issue. It helps to improve the user experience in the block editor and fix this issue. A little more help and the problem is certainly possible without sacrificing much performance. Certainly, a core developer is needed here who has more experience in overall performance degradation. But it's not unsolvable.

It should be noted that it is currently a bug and should be resolved. Because not every block may be nestable, groupable, or transformable.

There should be more freedom here to prevent blocks from being nestable, groupable, or transformable.

And currently there is absolutely no solution to do that.

**Comment by youknowriad** [comment]
Sure, I'd like to solve the problem too. Maybe we can start by creating the PR and checking the performance impact. If it's too much, we can also limit the solution to the other places to start with.

**Comment by CreativeDive** [comment]
@youknowriad Yes, it would be very interesting to see what difference these changes make in performance. I'm happy to pass this on to you :-)

**Comment by youknowriad** [comment]
Re-tested on trunk: a block whose parent is core/post-content can no longer be inserted (or grouped) inside a Group block; the root-only restriction holds. Closing as fixed.
